APCAppCore/APCAppCore/Library/Scoring/APCScoring.m

```diff
@@ -1054,58 +1312,108 @@ - (NSString *)lineGraph:(APCLineGraphView *) __unused graphView titleForXAxisAtI
 {
     NSDate *titleDate = nil;
     
-    titleDate = [[self.dataPoints objectAtIndex:pointIndex] valueForKey:kDatasetDateKey];
-
-    if (pointIndex == 0) {
-        [self.dateFormatter setDateFormat:@"MMM d"];
+    NSInteger numOfTitles = [self numberOfDivisionsInXAxisForLineGraph:graphView];
+    NSInteger actualIndex = ((self.dataPoints.count - 1)/numOfTitles + 1) * pointIndex;
+    
+    titleDate = [[self.dataPoints objectAtIndex:actualIndex] valueForKey:kDatasetDateKey];
+    
+    if (self.numberOfDays <= -(kNumberOfDaysIn3Months - 1)) {
+        // 3 Months or more
+        [self.dateFormatter setDateFormat:@"MMM"];
+    } else if (self.numberOfDays == -(kNumberOfDaysInMonth - 1)){
```

Comment:
This seems very error prone. Shouldn't we instead be using NSCalendar's various very helpful APIs for telling you which day is the start of each month, and cacheing the result if it is slow? Then this could be localized properly. As it is, this will have to be rewritten in order to localize (the templates should also ideally be generated by the NSDateFormatter dateFormatFromTemplate class method)